### DISCUSSION TITLE
Active balancing implementation

### DISCUSSION THREAD

**Discussion by MarcoZangrandi** [pybamm-team/PyBaMM#4422] [General]
Hello, I need your opinion on the following trends. 
I am implementing through PyBaMM a code that simulates the dynamics of battery packs. In particular I would like to implement active cell balancing, which I don't think liionpack offers yet. 
Currently I have simulated a parallel of 3 LFP cells (Prada + Chen2020 where the data was missing). 

I simulated a situation where the cells start from 3 different SOCs (100%, 90%, 80%) and a current of 0.001A (simulating a free response) is injected into the parallel to see how the cells rebalance naturally. I then implemented the balancing algorithm. 

This algorithm exchanges energy from the most charged cell to the most discharged cell and when the charges are very close together, it no longer allows energy exchanges. 
Below are the graphs of currents, terminal voltages, internal resistances and SOCs.
 
If someone has more experience than me on the subject, could you please explain to me whether it is possible that active balancing creates such effects, in particular on the terminal voltage? Because I would expect the terminal voltages to always be about the same (from Kirchhoff) as when the balancing is off, but maybe I am missing something. Also the internal resistances have strange trends but I don't think there's much I can do about that because it depends on voltages and currents.
If someone could give me his opinion, I would be very grateful, thanks.
  
![i_off](https://github.com/user-attachments/assets/a5b46c1d-603b-42b9-af08-cb515be3bd7f)
![i_on](https://github.com/user-attachments/assets/8a614344-bf70-462e-932a-e267c5ec3064)
![v_term_off](https://github.com/user-attachments/assets/9b343ccf-5099-4872-bbf7-5fc77c69ee68)
![v_term_on](https://github.com/user-attachments/assets/10ce9929-2eb4-4afa-8233-787bb706d645)
![r_off](https://github.com/user-attachments/assets/ccaa9212-5b19-4e97-8a1b-6f0c0415bb0b)
![r_on](https://github.com/user-attachments/assets/08dc2ef6-51bb-430b-a229-60ae96aa4970)
![soc](https://github.com/user-attachments/assets/0f1271fc-80a0-47f9-b950-b9e2e7e48044)

**Comment by Thiyaga2025** [comment]
hi @MarcoZangrandi ,

i am trying to do some pack modeling in pybamm as liionpack got some limitations. I am able to match the pack voltage by sequentially running pybamm cell models and summing up voltages. but not sure how to do cell balancing as you mentioned. can you please help with some idea in this regard.
thanks.

**Reply by MarcoZangrandi** [reply]
Hi, I have currently tried balancing a parallel group, but this does not reflect reality as active balancing is generally done for the series. 
For the series I have not implemented it yet.

I am having problems because the implementation I have done I think treats the 3 cells in parallel as separate systems, and so perturbations of currents correspond to different effects on voltages. This leads to not having equal terminal voltages. Which is not correct in my opinion. So I did not understand how to insert this constraint into the code.

The algorithm has a first part where the number of simultaneous exchanges (which defines the number of exchange pairs) and the distance at which I can exchange energy are fixed. More charged and more discharged cell pairs are then defined. Then the charge difference is calculated and divided by the sampling time, thus finding the current exchanged between the pairs. Once the branch current is found, it is given to the pybamm model. 
I hope that was helpful enough.

**Reply by Thiyaga2025** [reply]
@MarcoZangrandi ,

thanks a lot for the explanation.
